### DISCUSSION TITLE
Cannot get my if statements to work

### DISCUSSION THREAD

**Discussion by Mr-bkmtech** [micropython/micropython#11171] [RP2040 / Pico]
Hello

I am very new to Python.

I cannot seem to get my if statements to work.

```

import machine
from machine import Pin
import time

# define pins

RX = machine.Pin(1, machine.Pin.IN, Pin.PULL_DOWN)
CTR = machine.Pin(9, machine.Pin.IN, Pin.PULL_UP)
STAT = machine.Pin(10, machine.Pin.IN, Pin.PULL_UP)

TX = machine.Pin(0, machine.Pin.OUT)
# SDA = machine.Pin(2, machine.Pin.OUT)
# SCK = machine.Pin(3, machine.Pin.OUT)
A0 = machine.Pin(4, machine.Pin.OUT)
A1 = machine.Pin(5, machine.Pin.OUT)
A2 = machine.Pin(6, machine.Pin.OUT)
CS = machine.Pin(7, machine.Pin.OUT)
RH = machine.Pin(8, machine.Pin.OUT)
RESET = machine.Pin(11, machine.Pin.OUT)
HALT = machine.Pin(12, machine.Pin.OUT)
START = machine.Pin(13, machine.Pin.OUT)
LED = machine.Pin(25, machine.Pin.OUT)

# end define pins

while True:

    #   Create I2C object
    i2c = machine.I2C(1, scl=machine.Pin(3), sda=machine.Pin(2))

    #   Print out any addresses found
    devices = i2c.scan()

    if devices:
        for d in devices:
            print(hex(d))
            time.sleep(2)



# if STAT.value != ():
    LED(1)
    RX(1)
    #    SDA(1)
    #    SCK(1)
    A0(1)
    A1(1)
    A2(1)
    CS(1)
    RH(1)
    # CTR(1)
    # STAT(1)
    RESET(1)
    HALT(1)
    START(1)
    print("Stat Change")
    time.sleep(2)

# if CTR.value != ():
    LED(0)
    RX(0)
    #        SDA(0)
    #        SCK(0)
    A0(0)
    A1(0)
    A2(0)
    CS(0)
    RH(0)
    # CTR(0)
    # STAT(0)
    RESET(0)
    HALT(0)
    START(0)
    print("Ctr Change")
    time.sleep(3)

```

**Comment by glenn20** [comment]
You really should edit your message to fix up the formatting. It is impossible to be sure just what you are trying to do. (See the [github markdown docs](https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/creating-and-highlighting-code-blocks))

I have pasted below what it seems you meant. In that case, there is only one uncommented `if` statement (`if devices:`) and you haven't indicated how you know it is not working.

```python
import machine 
from machine import Pin 
import time

# define pins
RX = machine.Pin(1, machine.Pin.IN, Pin.PULL_DOWN) 
CTR = machine.Pin(9, machine.Pin.IN, Pin.PULL_UP) 
STAT = machine.Pin(10, machine.Pin.IN, Pin.PULL_UP)

TX = machine.Pin(0, machine.Pin.OUT)

# SDA = machine.Pin(2, machine.Pin.OUT)
# SCK = machine.Pin(3, machine.Pin.OUT)
A0 = machine.Pin(4, machine.Pin.OUT) 
A1 = machine.Pin(5, machine.Pin.OUT) 
A2 = machine.Pin(6, machine.Pin.OUT) 
CS = machine.Pin(7, machine.Pin.OUT) 
RH = machine.Pin(8, machine.Pin.OUT) 
RESET = machine.Pin(11, machine.Pin.OUT) 
HALT = machine.Pin(12, machine.Pin.OUT) 
START = machine.Pin(13, machine.Pin.OUT) 
LED = machine.Pin(25, machine.Pin.OUT)

# end define pins
while True:

    #   Create I2C object
    i2c = machine.I2C(1, scl=machine.Pin(3), sda=machine.Pin(2))

    #   Print out any addresses found
    devices = i2c.scan()

    if devices:
        for d in devices:
            print(hex(d))
            time.sleep(2)

# if STAT.value != ():
    LED(1)
    RX(1)
    #    SDA(1)
    #    SCK(1)
    A0(1)
    A1(1)
    A2(1)
    CS(1)
    RH(1)
    # CTR(1)
    # STAT(1)
    RESET(1)
    HALT(1)
    START(1)
    print("Stat Change")
    time.sleep(2)

# if CTR.value != ():
    LED(0)
    RX(0)
    #        SDA(0)
    #        SCK(0)
    A0(0)
    A1(0)
    A2(0)
    CS(0)
    RH(0)
    # CTR(0)
    # STAT(0)
    RESET(0)
    HALT(0)
    START(0)
    print("Ctr Change")
    time.sleep(3)
```

**Reply by Mr-bkmtech** [reply]
I will blame the formatting on GitHub, in fact,  I could not figure out how to correct it.

Actually, my if statements are working, partly.   But they are not recognizing a change in either CTR or STAT value.

I have double-checked my wiring and even checked the input values with a voltmeter.

Let's examine this one section of code.

STAT  is machine pin 10.   It is pulled up high, when I ground the input, all those IOs should go high, a logic 1

Is it written correctly?

```

# if STAT.value != ():
    LED(1)
    RX(1)
    #    SDA(1)
    #    SCK(1)
    A0(1)
    A1(1)
    A2(1)
    CS(1)
    RH(1)
    # CTR(1)
    # STAT(1)
    RESET(1)
    HALT(1)
    START(1)
    print("Stat Change")
    time.sleep(2)

```

**Reply by emersonsdnz** [reply]
Your first line, the if statement is commented out with the #. The # means everything on the line after it is ignored, therefore your code will run the same regardless of how the if statements are evaluated. Try removing them and see if it works?

**Comment by robert-hh** [comment]
> I will blame the formatting on GitHub, in fact, I could not figure out how to correct it.

Add lines with three backticks ```  before and after your code

**Reply by Mr-bkmtech** [reply]
Thanks

**Comment by Mr-bkmtech** [comment]
I have edited my code per Robert's direction and am still awaiting a reply.

**Reply by SergeiFranco** [reply]
Perhaps you should use STAT.value() opposed to STAT.value ?

https://docs.micropython.org/en/latest/library/machine.Pin.html

Edit: also the comparison is odd (unless I completely misunderstanding it):

The () in python is <class 'tuple'> so the STAT.value != () will always be true as STAT.value is <bound_method>.

If you compare the STAT.value() to (), then it also will not work, as you are comparing <class 'int'> to <class 'tuple'>:

The statement should be (in my opinion):

```
if STAT.value():
  # rest of the code
```

As STAT.value() (assuming it is a PIN) is either 0 or 1.

**Reply by Mr-bkmtech** [reply]
This must be so simple that I am overlooking it.

If you run the following code it does everything you would expect it to but for some reason, I cannot get the if statements to recognize the change in value on either CTR or STAT.

Is it a formatting or indentation error?

```


import machine
from machine import Pin
import time

# define pins

RX = machine.Pin(1, machine.Pin.IN, Pin.PULL_DOWN)
CTR = machine.Pin(9, machine.Pin.IN, Pin.PULL_UP)
STAT = machine.Pin(10, machine.Pin.IN, Pin.PULL_UP)

TX = machine.Pin(0, machine.Pin.OUT)
# SDA = machine.Pin(2, machine.Pin.OUT)
# SCK = machine.Pin(3, machine.Pin.OUT)
A0 = machine.Pin(4, machine.Pin.OUT)
A1 = machine.Pin(5, machine.Pin.OUT)
A2 = machine.Pin(6, machine.Pin.OUT)
CS = machine.Pin(7, machine.Pin.OUT)
RH = machine.Pin(8, machine.Pin.OUT)
RESET = machine.Pin(11, machine.Pin.OUT)
HALT = machine.Pin(12, machine.Pin.OUT)
START = machine.Pin(13, machine.Pin.OUT)
LED = machine.Pin(25, machine.Pin.OUT)

# end define pins

while True:

    #   Create I2C object
    i2c = machine.I2C(1, scl=machine.Pin(3), sda=machine.Pin(2))

    #   Print out any addresses found
    devices = i2c.scan()

    if devices:
        for d in devices:
            print(hex(d))
            time.sleep(1)

# if STAT.value():
    LED(1)
    RX(1)
    #    SDA(1)
    #    SCK(1)
    A0(1)
    A1(1)
    A2(1)
    CS(1)
    RH(1)
    # CTR(1)
    # STAT(1)
    RESET(1)
    HALT(1)
    START(1)
    print("Stat Change")
    # read and print the pin value
    print(STAT.value())
    time.sleep(1)

# if CTR.value():
    LED(0)
    RX(0)
    #        SDA(0)
    #        SCK(0)
    A0(0)
    A1(0)
    A2(0)
    CS(0)
    RH(0)
    # CTR(0)
    # STAT(0)
    RESET(0)
    HALT(0)
    START(0)
    print("Ctr Change")
    # read and print the pin value
    print(CTR.value())
    time.sleep(1)

```

**Reply by glenn20** [reply]
As written, this code just executes everything from the `while` to the end of the file in the loop. I think that is what you intended.

If you just delete the "# " before the two `if` statements, the scope of the while loop will end just **before** the first `if` statement, and the `if` statements will never be executed.

We would need to see the code (with **correct** indentation - please double check) after you restore the `if` statements to be sure exactly what you are doing.

**Comment by SergeiFranco** [comment]
If I were you,  I would try simple things first in the interactive shell.
I suspect you have indentation errors, if you uncomment the if statements from the above you will need to indent the inner section by exactly one indent.

By simple things I mean starting from one input pin that triggers one output pin inside of your while loop. Because you have a sleep(1) if your pin trigger happens while in the sleep(1) and trigger signal is shorter than 1 second it will miss it.

In anyway, it is best to use interrupts but at this stage I highly recommend to just try things out in the interactive mode (REPL):
https://docs.micropython.org/en/latest/esp8266/tutorial/repl.html
https://docs.micropython.org/en/latest/reference/repl.html


Once you have your pin -> led activation with your while loop (make it having sleep(0.1)), you should move on to interrupts ASAP:
https://docs.micropython.org/en/latest/esp8266/tutorial/pins.html#external-interrupts

Remember: indentation is crucial, and you should stick to 4 spaces (configure your editor/ide to convert the TABs to 4 spaces if you insist on "TAB"-ing).

**Reply by Mr-bkmtech** [reply]
I have made good progress with my programming.

I now have a problem installing this package using Thony IDE.

I am trying to install this package
https://gist.github.com/ScienceElectronicsFun/701828f27c857cf867b8b03c9b0ebab2

I get these errors
install install 'D:/Encrypted/BKM Engineering/Raspberry Pi/Programs and Codes/701828f27c857cf867b8b03c9b0ebab2-cd3acfa8cdf52d617c18362175a3c96fc637d960.zip'
Looking in indexes: http://127.0.0.1:36628
Processing d:\encrypted\bkm engineering\raspberry pi\programs and codes\701828f27c857cf867b8b03c9b0ebab2-cd3acfa8cdf52d617c18362175a3c96fc637d960.zip
ERROR: file:///D:/Encrypted/BKM%20Engineering/Raspberry%20Pi/Programs%20and%20Codes/701828f27c857cf867b8b03c9b0ebab2-cd3acfa8cdf52d617c18362175a3c96fc637d960.zip does not appear to be a Python project: neither 'setup.py' nor 'pyproject.toml' found.
Error Command '['C:\\Users\\pc\\AppData\\Local\\pipkin\\cache\\workspaces\\1c7bfecf12769f9897e393af77815ec0\\Scripts\\python.exe', '-I', '-m', 'pip', '--no-color', '--disable-pip-version-check', '--trusted-host', '127.0.0.1', 'install', '--no-compile', '--use-pep517', '--upgrade-strategy', 'only-if-needed', 'install', 'D:/Encrypted/BKM Engineering/Raspberry Pi/Programs and Codes/701828f27c857cf867b8b03c9b0ebab2-cd3acfa8cdf52d617c18362175a3c96fc637d960.zip', '--index-url', 'http://127.0.0.1:36628']' returned non-zero exit status 1.

**Comment by Mr-bkmtech** [comment]
I want to thank all of you for your help.

I am probably making a very basic mistake.
